bindings/rust/s2n-tls/src/cert_chain.rs

```diff
@@ -6,34 +6,160 @@ use s2n_tls_sys::*;
 use std::{
     marker::PhantomData,
     ptr::{self, NonNull},
+    sync::Arc,
 };
 
+/// Internal wrapper type used for a convenient drop implementation.
+///
+/// [CertificateChain] is internally reference counted. The reference counted `T`
+/// must have a drop implementation.
+struct CertificateChainHandle {
+    cert: NonNull<s2n_cert_chain_and_key>,
+    is_owned: bool,
+}
+
+// # Safety
+//
+// s2n_cert_chain_and_key objects can be sent across threads.
+unsafe impl Send for CertificateChainHandle {}
+unsafe impl Sync for CertificateChainHandle {}
+
+impl CertificateChainHandle {
+    fn from_owned(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: true,
+        }
+    }
+
+    fn from_reference(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: false,
+        }
+    }
+}
+
+impl Drop for CertificateChainHandle {
+    fn drop(&mut self) {
+        // ignore failures since there's not much we can do about it
+        if self.is_owned {
+            unsafe {
+                let _ = s2n_cert_chain_and_key_free(self.cert.as_ptr()).into_result();
+            }
+        }
+    }
+}
+
+pub struct Builder {
+    cert: CertificateChain<'static>,
+}
+
+impl Builder {
+    pub fn new() -> Result<Self, Error> {
+        Ok(Self {
+            cert: CertificateChain::allocate_owned()?,
+        })
+    }
+
+    /// Corresponds to [s2n_cert_chain_and_key_load_pem_bytes]
+    ///
+    /// This can be used with [crate::config::Builder::load_chain] to share a
+    /// single cert across multiple configs.
+    pub fn load_pem(&mut self, chain: &[u8], key: &[u8]) -> Result<&mut Self, Error> {
```

Comment:
I'm following our convention of builders returning `&mut Self`, but personally I like builders returning `Self` by value. Thoughts on whether it would be worth breaking from convention?